Codegen: Generate per-kind reference docs

undef1nd opened the pull request:
**What is this feature?**

Adds a jenny to generate docs for the core structured kind schemas.

Contributes to https://github.com/grafana/schematization-and-as-code-project/issues/28

Fixes https://github.com/grafana/schematization-and-as-code-project/issues/36 as per a2fd0b74b541fbeda54f70e6ce5f8e5e058e0b9a

Fixes https://github.com/grafana/schematization-and-as-code-project/issues/35 as per ceac10b01824d801f147467c428806f03b94bfc0

**Special notes for your reviewer**:
I added two new sections under `Developers` section: `Grafana schema` -> `Core kinds`

undef1nd committed:
Add docs generator

undef1nd committed:
Add json-to-md conversion

undef1nd committed:
Merge branch 'main' into codegen/docs

undef1nd committed:
Do not add generated header to the docs

undef1nd committed:
Fix path to generated docs

undef1nd committed:
Add error messages

undef1nd committed:
Linting

undef1nd committed:
Add draft attribute to the doc

undef1nd committed:
Update docs template

undef1nd committed:
Regenerate docs

undef1nd commented on docs/sources/developers/kinds/core/team/schema-reference.md:
```diff
@@ -0,0 +1,35 @@
+
+AccessControl metadata associated with a given resource.
+
+| Property | Type | Required | Description |
+|----------|------|----------|-------------|
+
```
Does not generate cue's `[string]: bool` properly: same issue as https://github.com/grafana/grafana/pull/59242#discussion_r1031579122

sdboyer replied:
Fix for this is now in #60496

undef1nd committed:
Add docs sections

undef1nd committed:
Merge branch 'main' into codegen/docs

undef1nd committed:
Regenerate docs

undef1nd committed:
Fix formatting

undef1nd committed:
Fix lint issues

undef1nd committed:
Merge branch 'main' into codegen/docs

sdboyer commented:
Given that @undef1nd has listed out discovered problems in a set of issues, as long as the generated docs only appear on `next`, then this is good enough to merge and iterate, from my perspective.

Would like @jdbaldry to give his sign off, too

undef1nd committed:
Merge branch 'main' into codegen/docs

undef1nd committed:
Remove check for kind type

undef1nd committed:
Disable prettier for generated docs

undef1nd commented on .prettierignore:
```diff
@@ -33,3 +33,6 @@ public/openapi3.json
 kinds/report.json
+
+# Generated schema docs
+docs/sources/developers/kinds/
```
I kept getting `prettier` warning reg table formatting. Disabling the prettier for the time being, till I find a better way to render tables in markdown (thus far github.com/olekukonko/tablewriter is being used for that).

This is my diff after running `yarn run prettier:write` on generated docs:
<img width="514" alt="Screenshot 2023-01-10 at 17 15 08" src="https://user-images.githubusercontent.com/10127682/211604102-29627a99-567b-4ede-bdfe-69d5561e03ff.png">

sdboyer replied:
Disabling linters on generated artifacts is a common and acceptable practice 👍🏻

undef1nd replied:
Thanks for confirming!

undef1nd committed:
Merge branch 'main' into codegen/docs

rhorvath commented on docs/sources/developers/kinds/_index.md:
```diff
@@ -0,0 +1,9 @@
+
+# Grafana schema
+
+[Core kinds]({{< relref "core/_index.md" >}})
```
Is the Core kinds link working for you? I had to remove the _index.md in my env:
`[Core kinds]({{< relref "core/" >}})`

joanlopez replied:
It works fine for me!

joanlopez committed:
Use schema ref names as identifiers for links & headers

joanlopez commented:
Fixed https://github.com/grafana/schematization-and-as-code-project/issues/36 with https://github.com/grafana/grafana/commit/a2fd0b74b541fbeda54f70e6ce5f8e5e058e0b9a (description updated!)

joanlopez commented on kinds/gen.go:
```diff
@@ -40,6 +40,7 @@ func main() {
 		codegen.BaseCoreRegistryJenny(filepath.Join("pkg", "registry", "corekind"), kindsys.GoCoreKindParentPath),
 		codegen.LatestMajorsOrXJenny(kindsys.TSCoreKindParentPath, codegen.TSTypesJenny{}),
 		codegen.TSVeneerIndexJenny(filepath.Join("packages", "grafana-schema", "src")),
+		codegen.DocsJenny(filepath.Join("docs", "sources", "developers", "kinds", "core"), true),
```
Should we do something similar for composable kinds? (I guess so, but wanna confirm) 🤔  cc/ @sdboyer

I guess would make sense to make `DocsJenny` responsible for both, so we could just pass the path til `.../kinds` and let if create `core` and `composable` directories. Any objection? @undef1nd

undef1nd replied:
That would be the best, agree.

sdboyer replied:
yep yep!

In an ideal world, the jenny itself only decides the filename, and there's a generic jenny pipeline component that we use when defining the overall pipeline to specify its leading path.

But the type system for jennies is too clunky right now to really support that, so what you're describing is :+1:. For now, what's really important is having just one jenny, regardless of whether this is a core or composable kind.

joanlopez committed:
Display the default value (if so) in the description

joanlopez commented:
Fixed https://github.com/grafana/schematization-and-as-code-project/issues/35 with ceac10b01824d801f147467c428806f03b94bfc0 (description updated!)

joanlopez committed:
Undo 'draft:false' introduced by mistake

joanlopez committed:
Merge remote-tracking branch 'upstream/main' into codegen/docs

undef1nd committed:
Merge branch 'main' into codegen/docs

jdbaldry commented on pkg/codegen/jenny_docs.go:
```diff
@@ -0,0 +1,486 @@
+}
+
+// -------------------- JSON to Markdown conversion --------------------
+// Copied from https://github.com/marcusolsson/json-schema-docs and sliggtly changed to fit the DocsJenny
```
nit: typo

```suggestion
// Copied from https://github.com/marcusolsson/json-schema-docs and slightly changed to fit the DocsJenny
```

undef1nd replied:
Thanks, Jack!

undef1nd committed:
Update pkg/codegen/jenny_docs.go

Co-authored-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Undraft and unlist kinds documentation  (#61476)

* Support running containers without root daemon

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* Use section shortcode to automatically list child pages

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* Undraft and unlist kinds documentation

This page and child pages are directly accessible but are not listed
in the table of contents.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* Add docs-preview to browse drafted pages

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

undef1nd committed:
Merge branch 'main' into codegen/docs

rhorvath committed:
Replace end of line and pipe characters in table codegen

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=60416) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=60416) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: undef1nd<br/>:white_check_mark: joanlopez<br/>:white_check_mark: jdbaldry<br/>:x: rhorvath<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=60416) it.</sub>

undef1nd committed:
Remove draft status from generated docs

undef1nd commented on docs/sources/developers/kinds/_index.md:
```diff
@@ -0,0 +1,10 @@
+title: Grafana schema
+weight: 200
+_build:
+  list: false
```
As discussed with @jdbaldry `draft: true` is not needed anymore so I removed it in this commit https://github.com/grafana/grafana/pull/60416/commits/967cab52fda006bed42e9fd59279a68f22c756a1

joanlopez reviewed (approved):
As agreed offline, I'd merge it as-is (looks fine enough for me) and work on the remaining issues later, in individually separated pull requests, easier to digest 🚀

undef1nd commented on pkg/codegen/jenny_docs.go:
```diff
@@ -0,0 +1,490 @@
+		return nil, fmt.Errorf("failed to marshal schema value to json: %v", err)
+	}
+
+	// We don't need entire json obj, only the value of components.schemas path
+	var obj struct {
+		Components struct {
+			Schemas json.RawMessage
+		}
+	}
+	err = json.Unmarshal(b, &obj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal schema json: %v", err)
+	}
+
+	// fixes the references between the types within a json after making components.schema.<types> the root of the json
+	kindJsonStr := strings.Replace(string(obj.Components.Schemas), "#/components/schemas/", "#/", -1)
```
Yep, right!
